File: Kotlin/Benchmark/src/main/kotlin/benchmark/fibonacci.kt
/* fibonacci.kt
 * Calculate Fibonacci numbers in the Kotlin
 * (http://kotlinlang.org) programming language
 */

package benchmark.fibonacci

import java.math.BigInteger

// naive implementation
fun fibNaive(n: Int): Int {
  if (n < 2) {
    return n
  } else {
    return (fibNaive(n-1) + fibNaive(n-2))
  }
}

// optimized tail-recursive implementation
fun fib(n: Int): BigInteger {
  tailrec fun loop(a: BigInteger, b: BigInteger, n: Int): BigInteger {
    return if (n == 0) a else loop(b, a + b, n - 1)
  }

  return loop(BigInteger.valueOf(0), BigInteger.valueOf(1), n)
}

File: Ruby/Benchmark/lib/Benchmark/fibonacci.rb
# fibonacci.rb ruby version to calculate Fibonacci series
#
# vim: ts=2 sw=2

module Benchmark
	module Fibonacci

		# naive implementation
		def Fibonacci.fib_naive(n)
			if n < 2 then
				n
			else
				Fibonacci.fib_naive(n-1) + Fibonacci.fib_naive(n-2)
			end
		end

		# tail-call optimized version
		def Fibonacci.fib(n)
			Fibonacci.fib_aux(n, 0, 1)
		end
		def Fibonacci.fib_aux(n, a, b)
			if n < 1 then
				a
			else
				Fibonacci.fib_aux(n-1, b, a+b)
			end
		end

	end
end


